analyzer/src/main/kotlin/managers/Yarn2.kt

```diff
@@ -126,21 +181,500 @@ class Yarn2(
         // We depend on a version >= 2, so we check the version for safety.
         definitionFiles.forEach { checkVersion(it.parentFile) }
 
-    override fun runInstall(workingDir: File): ProcessCapture =
-        // It should be enough for Yarn2+ to parse the lockfile to get the dependencies, and, if the former doesn't
-        // exist, run `yarn install --mode update-lockfile` to generate it (since yarn version 3.0.0-rc.10).
-        // Unfortunately, the lockfile doesn't contain the information required to build ORT's packages. An alternative
-        // could be to run `npm show --json` on each package present in the lockfile and rebuild a node_modules
-        // file hierarchy to be able to reuse the code present in the Npm package manager.
-        // Fortunately, an easier solution exist: by setting the option `nodeLinker` in the `yarnrc` file or as an
-        // environment variable, one can instruct Yarn to generate the node_modules files hierarchy.
-        run("install", workingDir = workingDir, environment = mapOf("YARN_NODE_LINKER" to "node-modules"))
-
-    override fun getRemotePackageDetails(workingDir: File, packageName: String): JsonNode {
-        // Yarn2+ 'info' command does not provide enough information so the default behavior of 'npm view --json' is
-        // kept. Note that it is mandatory to override this function as the super function calls `run()` which would
-        // call Yarn.
-        val process = delegateNpmAnalyzer.run(workingDir, "view", "--json", packageName)
-        return jsonMapper.readTree(process.stdout)
+    override fun resolveDependencies(definitionFile: File, labels: Map<String, String>): List<ProjectAnalyzerResult> {
+        val workingDir = definitionFile.parentFile
+
+        // Running `yarn install` before `yarn info` allows to get the real local package version. Otherwise, it will be
+        // a generic one such as '0.0.0-use.local'.
+
+        run("install", workingDir = workingDir)
+
+        // Query the list of all packages with their dependencies. The output is in NDJSON format.
+        val process = run(
+            "info",
+            "-A",
+            "-R",
+            "--manifest",
+            "--json",
+            workingDir = workingDir,
+            environment = mapOf("YARN_NODE_LINKER" to "pnp")
+        )
+
+        // First pass: Parse the headers and query each package details.
+        jsonMapper.createParser(process.stdout).use {
+            val iterator = jsonMapper.readValues<ObjectNode>(it)
+
+            queryPackageDetails(iterator, workingDir)
+        }
+
+        // Second pass: Parse the packages.
+        jsonMapper.createParser(process.stdout).use { parser ->
```

Comment:
Well there is a reason to it : if there was only one pass, it would be for each package:
1. parse the json from the header
2. query the details for this package
3. build the package

The problem would be that each package details would be queried one by one, with each time a new process spawn of `yarn npm info`.
The current implementation fires only (Npackages/BULK_DETAILS_SIZE) processes. Even if under the hood NPM still triggers one HTTP request per package, the whole execution is more performant.

Reply:
I was rather thinking of parsing the relevant information into temporary data structures once, which could then be iterated over as often as needed. But this means certainly more effort, so I am fine with keeping it as is.

Reply:
Ah yes I see what you meant. I didn't use temporary data structures on purpose to avoid keeping the full parsed JSON in memory. I wanted to stream the JSON file and parse and process one package at a time, to minimize memory usage.
Thanks for your understanding !